Implement create company and search

tedim52 opened the pull request:
-Implemented __create_company - via dictionary implementation and search functions in Google class
-Added test cases for each function
-Added resource.py module to hold Resource class
-Added __hash__, __eq__, __repr__ to Googler class
-Working now on testing and implementing distance function

tedim52 committed:
Implement create company and search

DeanAlvarez commented on Google.py:
```diff
@@ -78,9 +80,18 @@ def __create_company(self, googlers):
+        for googler in googlers:
+            # For every Googler, find their manager and map googler --> manager in dict
+            # NOTE: Believe this is implicitly O(n) making this algo O(n^2) :( if you're reading this guys, im using the thinker
+            manager = next((manager for manager in googlers if manager.id == googler.manager_id), None)
```
This is super terse. I'll be honest I don't really know what is going on here. 

That being said I don't think the map should be <googler,googler> I think it should be <integer,googler> you could build it in o(n) just by iterating through the csv file. Alternativenly, if you want to take in a list of googlers, just map the google Id to the googler. The idea is that if the manager_id in the data maps to a user_id you can follow the chain that way. We can talk about it tomorrow in the standup.

tedim52 replied:
I see what you're saying...  I'll make the change because then it does become O(1)

DeanAlvarez commented on Google.py:
```diff
@@ -18,13 +18,13 @@ class Google:
             googlers: A list of Googler objects.
         """
-        self.ceo = self.__create_company(googlers)
+        self.company = self.__create_company(googlers)
```
Why the two underscores? is that a style rule thing>

tedim52 replied:
https://www.geeksforgeeks.org/private-methods-in-python/ Yeah Python has no such thing as private methods like Java but double underscore indicates it should only be used within the class.. in this case, to create the dictionary company attribute.

DeanAlvarez replied:
cool, thanks for sharing!

DeanAlvarez commented on Google.py:
```diff
@@ -37,8 +37,10 @@ def search(self, googler_id):
         """
-        #TODO: Implement Googler Search function
-        return
+        googler = next((googler for googler in self.googlers if googler.id == googler_id), None)
```
If you implement the hashmap as user_id to googler this is o(1)

DeanAlvarez commented on Googler.py:
```diff
@@ -37,9 +37,28 @@ def __init__(self, attributes):
+        """
+        if isinstance(googler, type(self)):
+            return self.id == googler.id
+        return False
```
this is good.

DeanAlvarez commented on Googler.py:
```diff
@@ -37,9 +37,28 @@ def __init__(self, attributes):
         self.lowest_dir_id = attributes.get("lowest_dir_id")
         self.job_family = attributes.get("job_family")
-        self.resources = []
+        self.resources = [] #List of Resource objects from resource.py module
```
In Kayla's dictionary file that you might use to load in it is a set of resources not a list.

tedim52 replied:
Noted. We'll talk about the file format and decide the best way to ingest the information.

DeanAlvarez commented on resource.py:
```diff
@@ -0,0 +1,34 @@
+    # This technique is called cantor pairing
+    # See en.wikipedia.org/wiki/Pairing_function for more information
+    def __hash__(self):
+        return (((self.attr_one+self.attr_two+one)*(self.attr_one+self.attr_two)//two)+self.attr_two)
```
wow look at this absolutely EPIC hash function!

tedim52 replied:
made it myself :)

DeanAlvarez commented on Google.py:
```diff
@@ -78,9 +80,18 @@ def __create_company(self, googlers):
+        for googler in googlers:
+            # For every Googler, find their manager and map googler --> manager in dict
+            # NOTE: Believe this is implicitly O(n) making this algo O(n^2) :( if you're reading this guys, im using the thinker
+            manager = next((manager for manager in googlers if manager.id == googler.manager_id), None)
```
```suggestion
            company[googler.user_id] = googler
```

DeanAlvarez reviewed (commented):
We should talk about the implementation tomorrow morning

tedim52 committed:
Change implementation, implement distance function

tedim52 commented:
-Adjusted company implementation so dictionary maps <id:Googler>, search becomes O(1)
-Implemented basic distance function off dictionary implementation with test cases

tedim52 commented on Google.py:
```diff
@@ -55,8 +55,19 @@ def distance(self, googler_one, googler_two):
+        googler_two_managers = set()
+        while(len(googler_one_managers.intersection(googler_two_managers)) == 0):
+            if(googler_one.manager_id != -1):
+                googler_one_managers.add(googler_one.manager_id)
```
-1 indicates that the current googler is the "CEO" and has no manager so that the manager will only be added should they have one.

tedim52 replied:
Depends on how the googler_attribute_table.csv sets up the id to manager id relationship though.

DeanAlvarez reviewed (approved):
Looks good, speedy change!

tedim52 commented on Google_test.py:
```diff
@@ -3,22 +3,108 @@
+
+    def test_get_ceo(self):
+        # Test that the ceo has a None manager object
+        self.assertEqual(None, None)
```
Going to implement this test case when we receive updated dataset and I know how user id to manager id relationship will work.
